controllers/model/modelversion_controller.go

```diff
@@ -0,0 +1,456 @@
+		modelVersion.Status.ImageBuildPhase == modelv1alpha1.ImageBuildFailed {
+		r.Log.Info(fmt.Sprintf("image build %s", modelVersion.Status.ImageBuildPhase), "modelVersion", modelVersion.Name)
+		return reconcile.Result{}, nil
+	}
+
+	// Does the model exist, create it if not.
+	model := &modelv1alpha1.Model{}
+	err = r.Get(context.Background(), types.NamespacedName{Namespace: modelVersion.Namespace, Name: modelVersion.Spec.ModelName}, model)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.Info("model doesn't exist, create it.", "model", modelVersion.Spec.ModelName)
+			model := &modelv1alpha1.Model{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: modelVersion.Namespace,
+					Name:      modelVersion.Spec.ModelName,
+				},
+				Spec:   modelv1alpha1.ModelSpec{},
+				Status: modelv1alpha1.ModelStatus{},
+			}
+			model.Status.LatestVersion = &modelv1alpha1.VersionInfo{
+				ModelVersion: modelVersion.Name,
+			}
+			addModelToModelVersionOwnerReference(modelVersion, model)
+			err := r.Create(context.Background(), model)
+			if err != nil {
+				r.Log.Error(err, "failed to create model "+model.Name)
+				return reconcile.Result{Requeue: true}, err
+			}
+		} else {
+			return reconcile.Result{}, err
+		}
+	} else {
+		addModelToModelVersionOwnerReference(modelVersion, model)
+	}
+
+	pv := &v1.PersistentVolume{}
+	pvc := &v1.PersistentVolumeClaim{}
+
+	if modelVersion.Spec.Storage == nil {
+		r.Log.Error(err, "storage is undefined", "modelversion", modelVersion.Name)
+		return reconcile.Result{}, nil
+	} else {
+		err = r.createPVAndPVCForModelVersion(modelVersion, model, pv, pvc)
+		if err != nil {
+			r.Log.Error(err, "failed to create pv/pvc for modelversion", "modelversion", modelVersion.Name)
+			return reconcile.Result{Requeue: true}, err
+		}
+	}
+
+	// check if the pvc is bound
+	if pvc.Status.Phase != v1.ClaimBound {
+		// wait for the pv and pvc to be bound
+		r.Log.Info("waiting for pv and pvc to be bound", "modelversion", modelVersion.Name, "pv", pv.Name, "pvc", pvc.Name)
+		return ctrl.Result{Requeue: true, RequeueAfter: 1 * time.Second}, nil
+	}
+
+	modelVersionStatus := modelVersion.Status.DeepCopy()
+
+	// create kaniko pod to build container image\
+	// take the preceding 5 chars for image versionId
+	versionId := string(modelVersion.UID[:5])
+	imgBuildPodName := GetBuildImagePodName(model.Name, versionId)
+	imgBuildPod := &v1.Pod{}
+	err = r.Get(context.Background(), types.NamespacedName{Namespace: model.Namespace, Name: imgBuildPodName}, imgBuildPod)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			modelVersionStatus.Image = modelVersion.Spec.ImageRepo + ":v" + versionId
+			modelVersionStatus.ImageBuildPhase = modelv1alpha1.ImageBuilding
+			modelVersionStatus.Message = "Image building started."
+			imgBuildPod = createImgBuildPod(modelVersion, pvc, imgBuildPodName, modelVersionStatus.Image)
+			err = r.Create(context.Background(), imgBuildPod)
+			if err != nil {
```

Comment:
if the err is AlreadyExists, it can be continued with the following steps

Reply:
this  if condition will enter when imgBuildPod does not exist